enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/PropertyExistenceEnforcer.java

```diff
@@ -42,191 +49,406 @@
-        Cursor<PropertyItem> cursor = storeStatement().acquirePropertyCursor( node.nextPropertyId(), lock );
-        return txState.augmentPropertyCursor( cursor, txState.getNodeState( node.id() ) );
+        for ( int key : requiredKeys )
+        {
+            if ( !propertyKeyIds.contains( key ) )
+            {
+                failNode( id, label, key );
+            }
+        }
     }
 
-    private StorageStatement storeStatement()
+    private void failNode( long id, int label, int propertyKey )
+            throws NodePropertyExistenceException
     {
-        return storageStatement == null ? storageStatement = storeLayer.newStatement() : storageStatement;
+        for ( LabelSchemaDescriptor constraint : nodeConstraints )
+        {
+            if ( constraint.getLabelId() == label && contains( constraint.getPropertyIds(), propertyKey ) )
+            {
+                throw new NodePropertyExistenceException( constraint, VALIDATION, id );
+            }
+        }
+        throw new IllegalStateException( format(
+                "Node constraint for label=%d, propertyKey=%d should exist.",
+                label, propertyKey ) );
     }
 
-    @Override
-    public void close()
+    private void failRelationship( long id, int relationshipType, int propertyKey )
+            throws RelationshipPropertyExistenceException
     {
-        super.close();
-        if ( storageStatement != null )
+        for ( RelationTypeSchemaDescriptor constraint : relationshipConstraints )
         {
-            storageStatement.close();
+            if ( constraint.getRelTypeId() == relationshipType && contains( constraint.getPropertyIds(), propertyKey ) )
+            {
+                throw new RelationshipPropertyExistenceException( constraint, VALIDATION, id );
+            }
         }
+        throw new IllegalStateException( format(
+                "Relationship constraint for relationshipType=%d, propertyKey=%d should exist.",
+                relationshipType, propertyKey ) );
     }
 
-    private void validateRelationship( long id ) throws RelationshipPropertyExistenceException
+    private static void update( PrimitiveIntObjectMap<int[]> map, int key, int[] values )
     {
-        if ( relTypeExistenceConstraints.isEmpty() )
+        Arrays.sort( values );
+        int[] current = map.get( key );
+        if ( current != null )
         {
-            return;
+            values = merge( current, values );
         }
+        map.put( key, values );
+    }
 
-        try ( Cursor<RelationshipItem> relationship = relationshipCursor( id ) )
+    /**
+     * Merges two sets of integers represented as sorted arrays.
+     *
+     * @param lhs
+     *         a set of integers, represented as a sorted array.
+     * @param rhs
+     *         a set of integers, represented as a sorted array.
+     * @return a set of integers, represented as a sorted array.
+     */
+    static int[] merge( int[] lhs, int[] rhs )
```

Comment:
Since this method has nothing to do with property existence per se, how about lifting this method to some util class? also `contains`. New class `org.neo4j.collection.primitive.PrimitiveSortedArrays`?